### PR TITLE
Simplify CPU kernel launch

### LINKED ISSUE #1492: Create perfectly forwarding function objects for CPU back-ends
The `TaskKernel` implementations for the TBB, OpenMP 2.x and Serial back-ends feature the following comment:

```c++
// TODO: With C++14 we could create a perfectly argument forwarding function object within the constructor.
```

We should address this as part of the C++17 migration.

### PR DESCRIPTION
**Pull request by j-stephan**
Fixes #1492 (with an even simpler solution).

### CODE REVIEW

File: include/alpaka/kernel/TaskKernelCpuOmp2Blocks.hpp [hunk 1]
```diff
@@ -910,7 +903,7 @@ namespace alpaka
 #    endif
                 acc.m_gridBlockIdx = mapIdx<TDim::value>(index, gridBlockExtent);
 
-                boundKernelFnObj(acc);
+                std::apply(m_kernelFnObj, std::tuple_cat(std::make_tuple(acc), m_args));
```
Comment (bernhardmgruber) [inline include/alpaka/kernel/TaskKernelCpuOmp2Blocks.hpp]: I wonder whether that changes the semantic in a problematic way. Your new solution makes N threads access `this` to fetch `m_kernelFnObj` and `m_args` in parallel. All kernel threads also share the functor and argument instances. This is definitely a problem if the functor mutates itself. A user could also now get a reference to one of `m_args`, if they use a mutable reference in the kernel function interface and produce a race condition by writing to it.

I think I would prefer giving each thread an independent copy of the functor and args.

Comment (j-stephan) [reply include/alpaka/kernel/TaskKernelCpuOmp2Blocks.hpp]: > This is definitely a problem if the functor mutates itself.

I don't think we allow this as it should be marked `const`.

Comment (psychocoderHPC) [reply include/alpaka/kernel/TaskKernelCpuOmp2Blocks.hpp]: Yes arguments are const, a non-const reference in the user kernel signature should not work. An open question for OpenMP is if it is better to use variables `shared` or an instance per thread. I would say it does not matter if all arguments are const but I am not fully sure.

Comment (bernhardmgruber) [reply include/alpaka/kernel/TaskKernelCpuOmp2Blocks.hpp]: Ok. That solves the issue for `m_args`. What if the `operator()` on the kernel functor is not const? Is that allowed? If not, we are fine with the changes.

Comment (j-stephan) [reply include/alpaka/kernel/TaskKernelCpuOmp2Blocks.hpp]: I don't think non-`const` function objects are allowed. Every instance of the kernel `operator()` I've seen so far has been marked as `const`.

Comment (j-stephan) [reply include/alpaka/kernel/TaskKernelCpuOmp2Blocks.hpp]: Our docs explicitly say that the `operator()` must be `const`. See the code snippet at https://alpaka.readthedocs.io/en/latest/basic/library.html#kernel-definition

Comment (bernhardmgruber) [reply include/alpaka/kernel/TaskKernelCpuOmp2Blocks.hpp]: Then your changes are fine!

**Review by bernhardmgruber — DISMISSED** [review]

**Review by psychocoderHPC — APPROVED** [review]

**Pull request merged**